File: 2022/day03-1.ts
function solution(content: string) {
  return content
    .split("\n")
    .map((item) => priority(repeated(item)))
    .reduce((acc, cur) => acc + cur, 0);
}

function repeated(line: string) {
  const fst = line.slice(0, line.length / 2);
  const snd = line.slice(line.length / 2);
  return fst.split("").find((item) => snd.includes(item))!;
}

function priority(ch: string) {
  if (ch[0] >= "a" && ch[0] <= "z")
    return ch.charCodeAt(0) - "a".charCodeAt(0) + 1;
  if (ch[0] >= "A" && ch[0] <= "Z")
    return ch.charCodeAt(0) - "A".charCodeAt(0) + 27;
  return 0;
}

Deno.readTextFile("day03.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);

File: 2022/day03_1.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from typing import Any


def solution(content: list[str]) -> Any:
    return sum(priority(repeated(line)) for line in content)


def repeated(line: str):
    fst = line[: len(line) // 2]
    snd = line[len(line) // 2 :]
    return next(ch for ch in fst if ch in snd)


def priority(ch: str):
    if "a" <= ch[0] <= "z":
        return ord(ch) - ord("a") + 1
    if "A" <= ch[0] <= "Z":
        return ord(ch) - ord("A") + 27
    return 0


def main() -> None:
    with open("day03.dat", "r", encoding="utf8") as file:
        print(solution(file.readlines()))


if __name__ == "__main__":
    main()
